ArchiSteamFarm/Statistics.cs

```diff
@@ -457,22 +459,54 @@ internal sealed class Statistics : IDisposable {
 						do {
 							match = false;
 
-							foreach ((ulong ourItem, uint ourAmount) in ourFullSet.Where(item => item.Value > 1).OrderByDescending(item => item.Value)) {
-								if (!ourTradableSet.TryGetValue(ourItem, out uint tradableAmount) || (tradableAmount == 0)) {
+							foreach ((ulong ourItem, uint ourFullAmount) in ourFullSet.Where(item => item.Value > 1).OrderByDescending(item => item.Value)) {
+								if (!ourTradableSet.TryGetValue(ourItem, out uint ourTradableAmount) || (ourTradableAmount == 0)) {
 									continue;
 								}
 
-								foreach ((ulong theirItem, _) in theirItems.OrderBy(item => ourFullSet.TryGetValue(item.Key, out uint ourAmountOfTheirItem) ? ourAmountOfTheirItem : 0)) {
-									if (ourFullSet.TryGetValue(theirItem, out uint ourAmountOfTheirItem) && (ourAmount <= ourAmountOfTheirItem + 1)) {
+								foreach ((ulong theirItem, uint theirTradableAmount) in theirTradableItems.OrderBy(item => ourFullSet.TryGetValue(item.Key, out uint ourAmountOfTheirItem) ? ourAmountOfTheirItem : 0)) {
+									if (ourFullSet.TryGetValue(theirItem, out uint ourAmountOfTheirItem) && (ourFullAmount <= ourAmountOfTheirItem + 1)) {
 										continue;
 									}
 
+									if (!listedUser.MatchEverything) {
+										// We have a potential match, let's check fairness for them
+										fairClassIDsToGive.TryGetValue(ourItem, out uint fairGivenAmount);
+										fairClassIDsToReceive.TryGetValue(theirItem, out uint fairReceivedAmount);
+										fairClassIDsToGive[ourItem] = ++fairGivenAmount;
+										fairClassIDsToReceive[theirItem] = ++fairReceivedAmount;
+										
+										// Filter their inventory for the sets we're trading or have traded with this user
+										HashSet<Steam.Asset> fairFiltered = new HashSet<Steam.Asset>(theirInventory.Where(item => (item.RealAppID == set.RealAppID) && (item.Type == set.Type) && (item.Rarity == set.Rarity) || skippedSetsThisTrade.Any(skippedSets => skippedSets.RealAppID == item.RealAppID && skippedSets.Type == item.Type && skippedSets.Rarity == item.Rarity)).Select(item => item.CreateShallowCopy()).ToHashSet());
```

Comment:
You're creating hashset twice here, remove `new` constructor, `ToHashSet()` already returns you one.